File: vm.c

#include <stdio.h>
#include <stdlib.h>
#include <memory.h>
#include "ci.h"

extern int * sym;

extern int * be;
extern char * bd;

const char *op_codes =
    "LEA ,IMM ,JMP ,JSR ,BZ  ,BNZ ,ENT ,ADJ ,LGB ,"
    "LEV ,"
    "LI  ,LC  ,SI  ,SC  ,PSH ,"

    "OR  ,XOR ,AND ,EQ  ,NE  ,LT  ,GT  ,LE  ,GE  ,"
    "SHL ,SHR ,ADD ,SUB ,MUL ,DIV ,MOD ,"

    "OPEN,READ,CLOS,PRTF,MALC,MSET,MCMP,EXIT,";

static void
debug_info(int *pc, int i, int cycle) {
    printf("%d> %d: %.4s", cycle, (int)pc, &op_codes[i * 5]);
    if (i <= LGB) {
        printf(" %d\n", *pc);
    } else {
        printf("\n");
    }
}

int run_c(int argc, char **argv, int debug) {
    int *pc, *sp, *bp = NULL, a = 0, cycle; // vm registers

    int *id = sym;
    while (id[Tk]) {
        if (!memcmp((const void *)id[Name], "main", 4)) {
            break;
        }
        id = id + Idsz;
    }

    if (!(pc = be + id[Val])) {
        printf("main() not defined\n");
        return -1;
    }

    int stack_size = 128 * 1024;
    if (!(sp = malloc(stack_size * sizeof(int)))) {
        printf("could not malloc(%d) stack area\n", stack_size);
        return -1;
    }

    int i, *t; // temps
    // setup stack
    sp    = sp + stack_size;
    *--sp = EXIT; // call exit if main returns
    *--sp = PSH;
    t     = sp;
    *--sp = argc;
    *--sp = (int)argv;
    *--sp = (int)t;

    cycle = 0;

#define ci_dispatch(o)  switch(o)
#define ci_case(c,b)    case c: {b} break;
#define ci_default(b)   default: {b};

    while (1) {
        i = *pc++; ++cycle;
        if (debug) debug_info(pc, i, cycle);

        ci_dispatch(i) {
            ci_case(LEA, a = (int)(bp + *pc++);)                     // load local address
            ci_case(LGB, a = (int)(bd + *pc++);)                     // load global address
            ci_case(IMM, a = *pc++;)                                 // load immediate
            ci_case(JMP, pc = pc + *pc;)                             // jump
            ci_case(JSR, *--sp = (int)(pc + 1); pc = be + *pc;)      // jump to subroutine
            ci_case(BZ,  pc = a ? pc + 1 : pc + *pc;)                // branch if zero
            ci_case(BNZ, pc = a ? pc + *pc : pc + 1;)                // branch if not zero
            ci_case(ENT, *--sp = (int)bp; bp = sp; sp = sp - *pc++;) // enter subroutine
            ci_case(ADJ, sp = sp + *pc++;)                           // stack adjust

            ci_case(LEV, sp = bp; bp = (int *)*sp++; pc = (int *)*sp++;) // leave subroutine

            ci_case(LI,  a = *(int *)a;)          // load int
            ci_case(LC,  a = *(char *)a;)         // load char
            ci_case(SI,  *(int *)*sp++ = a;)      // store int
            ci_case(SC,  a = *(char *)*sp++ = a;) // store char
            ci_case(PSH, *--sp = a;)              // push

            ci_case(OR,  a = *sp++ |  a;)
            ci_case(XOR, a = *sp++ ^  a;)
            ci_case(AND, a = *sp++ &  a;)
            ci_case(EQ,  a = *sp++ == a;)
            ci_case(NE,  a = *sp++ != a;)
            ci_case(LT,  a = *sp++ <  a;)
            ci_case(GT,  a = *sp++ >  a;)
            ci_case(LE,  a = *sp++ <= a;)
            ci_case(GE,  a = *sp++ >= a;)
            ci_case(SHL, a = *sp++ << a;)
            ci_case(SHR, a = *sp++ >> a;)
            ci_case(ADD, a = *sp++ +  a;)
            ci_case(SUB, a = *sp++ -  a;)
            ci_case(MUL, a = *sp++ *  a;)
            ci_case(DIV, a = *sp++ /  a;)
            ci_case(MOD, a = *sp++ %  a;)

            ci_case(OPEN, a = (int)fopen((const char *)sp[1], (const char *)*sp);)
            ci_case(READ, a = fread((char *)sp[3], sp[2], sp[1], (FILE *)*sp);)
            ci_case(CLOS, a = fclose((FILE *)*sp);)
            ci_case(PRTF,
                    t = sp + pc[1];
                    a = printf((char *)t[-1], t[-2], t[-3], t[-4], t[-5], t[-6]);
                    )
            ci_case(MALC, a = (int)malloc(*sp);)
            ci_case(MSET, a = (int)memset((char *)sp[2], sp[1], *sp);)
            ci_case(MCMP, a = memcmp((char *)sp[2], (char *)sp[1], *sp);)
            ci_case(EXIT,
                    printf("exit(%d) cycle = %d\n", *sp, cycle);
                    return *sp;)

            ci_default(
                    printf("unknown instruction = %d! cycle = %d\n", i, cycle);
                    return -1;)
        }
    }
    return 0;
}

File: vm.lua
#!/usr/bin/env lua

local LEA = 0
local IMM = 1
local JMP = 2
local JSR = 3
local BZ = 4
local BNZ = 5
local ENT = 6
local ADJ = 7
local LGB = 8
local LEV = 9

local LI = 10
local LC = 11
local SI = 12
local SC = 13
local PSH = 14

local OR = 15
local XOR = 16
local AND = 17
local EQ = 18
local NE = 19
local LT = 20
local GT = 21
local LE = 22
local GE = 23

local SHL = 24
local SHR = 25
local ADD = 26
local SUB = 27
local MUL = 28
local DIV = 29
local MOD = 30

local OPEN = 31
local READ = 32
local CLOS = 33
local PRTF = 34
local MALC = 35
local MSET = 36
local MCMP = 37
local EXIT = 38

--- module vm
-- Thu 19:11 May 12

-- {
--  main
--  text {}
--  data ''
-- }
local function lc_load(fn)
    local f = io.open(fn, 'rb')
    if not f then return nil end

    local hd = assert(f:read(12))

    local main_addr, t_sz, d_sz = ('<i4i4i4'):unpack(hd)

    local text = {}
    local r = {main = main_addr, text = text}

    for i = 1, t_sz // 4 do
        local code = ('<i4'):unpack(f:read(4))
        table.insert(text, code)
    end
    r.data = f:read(d_sz)
    f:close()
    return r
end


-- 封装栈的操作
local function new_stack()
    return {
        sp = 0,
        push = function(self, value)
            self.sp = self.sp + 1
            table.insert(self, value)
        end,
        pop = function(self)
            self.sp = self.sp - 1
            return table.remove(self)
        end,
        get_sp = function(self) return self.sp end,
        set_sp = function(self, value) self.sp = value end
    }
end

-- 封装 PC寄存器的操作？ 为何叫text?
local function new_text(p)
    return {
        data = p.text,
        pc = p.main,
        get_pc_inc = function(self)
            local i = self.pc
            self.pc = i + 1
            return self.data[i]
        end,
        current_pc = function(self) return self.pc end,
        current_pc_value = function(self) return self.data[self.pc] end,
        add_pc_by = function(self, value) self.pc = self.pc + value end,
        set_pc = function(self, value) self.pc = value end
    }
end

local function lc_run(p)
    assert(p.main and p.text and p.data)
    local data = p.data

    local sp = new_stack() -- 与C版不同，栈的增长方向向上
    sp:push(EXIT)
    sp:push(PSH)
    local t = sp:get_sp()

    -- TODO
    local argc, argv = 0, 0

    sp:push(argc)
    sp:push(argv)
    sp:push(t)

    local cycle = 0
    local bp = 'NULL'
    local pc = new_text(p)
    local a = 0
    while true do
        local i = pc:get_pc_inc()
        cycle = cycle + 1
        if true then print(pc:current_pc(), i, cycle) end

        if cycle > 100 then break end

        if i == 0 then -- LEA
            a = bp + pc:get_pc_inc()
        elseif i == 1 then -- IMM
            a = pc:get_pc_inc()
        elseif i == 2 then -- JMP
            pc:add_pc_by(pc:current_pc_value())
        elseif i == 3 then -- JSR
            sp:push(pc:current_pc() + 1)
            pc:set_pc(pc:current_pc_value())
        elseif i == 4 then -- BZ
            if a ~= 0 then
                pc:set_pc(pc:current_pc() + 1)
            else
                pc:set_pc(pc:current_pc() + pc:current_pc_value())
            end
        elseif i == 5 then -- BNZ
            if a ~= 0 then
                pc:set_pc(pc:current_pc() + pc:current_pc_value())
            else
                pc:set_pc(pc:current_pc() + 1)
            end
        elseif i == 6 then -- ENT
            sp:push(bp)
            bp = sp:get_sp()
            sp:set_sp(sp:get_sp() - pc:get_pc_inc())
        elseif i == 7 then -- ADJ
            sp:set_sp(sp:get_sp() + pc:get_pc_inc())
        elseif i == 8 then -- LGB
            a = pc:get_pc_inc()
        elseif i == 9 then -- LEV

        elseif i == 10 then -- LI
            -- Lua中怎么实现寻址这样的操作呢
            -- Sun 16:09 May 15
        elseif i == 11 then -- LC
        elseif i == 12 then -- SI
        elseif i == 13 then -- SC
        elseif i == 14 then -- PSH

        elseif i == 15 then -- OR
        elseif i == 16 then -- XOR
        elseif i == 17 then -- AND
        elseif i == 18 then -- EQ
        elseif i == 19 then -- NE
        elseif i == 20 then -- LT
        elseif i == 21 then -- GT
        elseif i == 22 then -- LE
        elseif i == 23 then -- GE

        elseif i == 24 then -- SHL
        elseif i == 25 then -- SHR
        elseif i == 26 then -- ADD
        elseif i == 27 then -- SUB
        elseif i == 28 then -- MUL
        elseif i == 29 then -- DIV
        elseif i == 30 then -- MOD

        elseif i == 31 then -- OPEN
        elseif i == 32 then -- READ
        elseif i == 33 then -- CLOS
        elseif i == 34 then -- PRTF
        elseif i == 35 then -- MALC
        elseif i == 36 then -- MSET
        elseif i == 37 then -- MCMP
        elseif i == 38 then -- EXIT
        end
    end
end

local function lc_dump(p)
    for i = 1, #p.text do print(p.text[i]) end
    print(p.main, p.data)
end

local function main()
    local fn = arg[1]
    if fn then
        local p = lc_load(fn)
        -- lc_dump(p)
        lc_run(p)
    end
end

return main()
